### PR TITLE
[react] better types for React.Children, second generic parameter for ReactElement

### PR DESCRIPTION
**Pull request by ferdaber**
### Context
A lot of these changes are to help library authors better manipulate the `ReactElement` black box, and children in general.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
If changing an existing definition:
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -2670,7 +2671,7 @@ type ReactManagedAttributes<C, P> = C extends { propTypes: infer T; defaultProps
 declare global {
     namespace JSX {
         // tslint:disable-next-line:no-empty-interface
-        interface Element extends React.ReactElement<any> { }
+        interface Element extends React.ReactElement<any, any> { }
```
Comment (weswigham) [inline types/react/index.d.ts]: Maybe just let the second arg be `default`'d to the constraint, rather than `any` here.

**Comment by uniqueiniquity** [comment]
@Jessidhia @johnnyreilly Can you review?

**Comment by johnnyreilly** [comment]
Merged - thanks!

**Comment by yang** [comment]
This seems to cause React.Children.only() on a ReactNode to simply return ReactNode (which may be an array of stuff), rather than a singular ReactElement.  React.Children.only is supposed to be a runtime mechanism for ensuring some input - which may or may not be an array - turns into a single element.

Recall:

```
    type ReactText = string | number;
    type ReactChild = ReactElement | ReactText;

    interface ReactNodeArray extends Array<ReactNode> {}
    type ReactFragment = {} | ReactNodeArray;
    type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null | undefined;
```

**Comment by yang** [comment]
@ferdaber @johnnyreilly This seems to cause React.Children.only() on a ReactNode to simply return ReactNode (which may be an array of stuff), rather than a singular ReactElement. React.Children.only is supposed to be a runtime mechanism for ensuring some input - which may or may not be an array - turns into a single element.

Some context:

```
    type ReactText = string | number;
    type ReactChild = ReactElement | ReactText;

    interface ReactNodeArray extends Array<ReactNode> {}
    type ReactFragment = {} | ReactNodeArray;
    type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null | undefined;
```

**Comment by yang** [comment]
Oops, disregard - I misread the inferred types.

**Pull request merged**